### ISSUE TITLE
Users get sent to /_fos_user_context_hash upon login (ezpublish application)

### ISSUE THREAD

**Issue by gggeek** [FriendsOfSymfony/FOSHttpCache#287]
Not sure if it is a bug in FOSHttpCache, eZPublish or custom code.
The observed behaviour (no varnish, http cache on, ezp 5.4.5):
- user accesses the site
- gets redirected to /login
- fills in the form, posts to /login_check
- gets redirected to /_fos_user_context_hash

Since I fixed this by setting:

security:
    firewalls:
        ezpublish_front:
            form_login:
                always_use_default_target_path: true

I suspect that what goes on here is that the internal request made by Sf to retrieve the user-hash gets somehow 'remember' by the session as the last url visited by the end user just before logging in.

Any tips on fixing this in a more flexible way? Or is this really a bug?

**Comment by dbu** [comment]
is /_fos_user_context_hash under a firewall? that would be wrong. if it is not under a firewall, the request to the actually requested page should be after the hash lookup, and trigger the redirect to login, so the last access would be on the actually requested page.

having the hash url access protected is not an intended scenario. if the user is not logged in, that is also a context hash.

**Comment by gggeek** [comment]
@dbu the /_fos_user_context_hash is indeed under a firewall section, but it has "anonymous: ~" set.
Do you think that it should be added to a separate firewall with "security: false" instead ?
On a side note: are there any security implications into allowing /_fos_user_context_hash to be accessible by end users? Should we block access it at the reverse proxy (varnish) ?

**Comment by gggeek** [comment]
ping @andrerom @lolautruche

**Comment by andrerom** [comment]
@ggeek be aware we have blocked 1.2.7 and higher as of 5.4.6. Are you on any of those FOS versions?

**Comment by stof** [comment]
@gggeek @dbu  it cannot be under a separate firewall, as this is precisely about computing a hash based on the logged in user, and so needs the firewall to run to determine the logged in user.

What is needed is to prevent Symfony from storing this URL as the requested URL in the session.
See http://symfony.com/doc/current/cookbook/security/target_path.html for the way to exclude it from being stored (the example in the doc detects AJAX requests, which is now useless as the builtin logic handles it, but you can just put ``if ('user_context_hash' === $request->attributes->get('_route')) {`` there (adapt the route name to your needs)

**Comment by gggeek** [comment]
@andrerom I have seen the problem happening on both a community version of eZ (foshttpcache: 1.4.1, foshc-bundle: 1.3.6) and an enterprise version 5.4.6 (foshttpcache: 1.4.2, foshc-bundle: 1.3.7). Are you sure that the lockdown is in eZP 5.4.6 ?

**Comment by gggeek** [comment]
@stof thanks for chiming in. Do you reckon that the fix you mention could be added to the cacheBudle itself?

**Comment by gggeek** [comment]
PS: @andrerom I guess you meant "1.3.7 and higher", as there is no version 1.2.7 ?
If so, it seems that this is a separate bug, as it happened to us even on 1.3.6...

**Comment by ddeboer** [comment]
Just to clear up some confusion: `/_fos_user_context_hash` definitely needs to be inside a firewall. We can make the [note in the docs](http://foshttpcachebundle.readthedocs.io/en/latest/features/user-context.html#configuration) on this clearer though: “… and make sure that this path is covered by your firewall configuration.”

I agree with @stof’s solution. It’s unfortunate that we have to override ExceptionListener, but I see no other way. @dbu Do you agree?

**Comment by andrerom** [comment]
> PS: @andrerom I guess you meant "1.3.7 and higher", as there is no version 1.2.7 ?
> If so, it seems that this is a separate bug, as it happened to us even on 1.3.6...

yes and yes, updated comment now that I'm not on phone :)

**Comment by dbu** [comment]
but why would the fos_user_context_hash route ever run into a security exception? it should return the hash for the anonymous user if there is no user credentials.

we can add a security exception listener in this bundle that is active by default and helps fixing a borked setup a bit. but i don't think things will work out as expected if something is misconfigured.

the example in the symfony doc should be updated by the way. its not only that the example is not relevant anymore but also that its using the class parameters which have been dropped in symfony 3 afaik.

**Comment by dbu** [comment]
btw, with the correct setup, i would expect there to be some issues with this feature. if /my/page redirects to login for anonymous users, we might have that cached - unless the security failure leads to making the page not cached. then every direct hit creates a new session and stores the url and happens to work out as the login is individual as well.

**Comment by gggeek** [comment]
@dbu 
1. if I understand correctly, the security exception listener is only needed in cases where there is misconfiguration. In my case, I have no idea what the misconfiguration might be. I'd be happy to fix it the proper way if someone gave a tip on how to identity it.
2. I am sorry but your last comment is too complex for me: is "this feature" is the exception listener? and the issue is that /my/page would be cached for anon users, or that it would not be cached?

**Comment by dbu** [comment]
@gggeek 

1. maybe try to "fake" the request with curl on the command line to see what you get back? there is a special accept-encoding header that varnish is sending, so just opening in a browser won't work. and look at the symfony logfile, having set log level to debug.

2. sorry that was a bit unclear. this feature meant "after login, redirect the user to the page he originally wanted to go to". if /my/page is in the firewall and thus triggers a redirect, if that redirect response is cached by varnish, redirecting the user after the login will not work because symfony never saw the request from that user. if each anonymous access to /my/page generates a hit on the backend however, you would get sessions that you reuse on /login and then do the redirect. if its a direct hit, this really should be the case as there should be a set-cookie header which makes the response un-cacheable.

**Comment by gggeek** [comment]
1. just to be clear: the spurious redirection happens in my case for 2 installations; one does use varnish, the other one does not, and it use the sf internal reverse proxy (http cache is enabled).

2. I would imagine that the redirect response which is generated by Sf when the anon user first tries to go to /my/page is generated with a no-cache header as well as 302 status, so that it does never get cached.

**Comment by dbu** [comment]
1. the php reverse proxy also uses that special header. try to simulate a request to that url to see why it fails. it should return a hash, even for anonymous

2. likely true, yes.

**Comment by gggeek** [comment]
about 1: my problem is not what is generated by the url /_fos_user_context_hash (at least, not that I know of), it is that the users browser gets redirected to it, which should not happen

**Comment by dbu** [comment]
i think the reason for that problem is what happens during the request to /_fos_user_context_hash. the only reason why the user gets redirected there is when that url triggers a login request and is stored in the users session as the page they wanted to see. if it would just generate a hash, varnish would next try to access the actual page with that hash, and *then* get the redirection request to login, and symfony would store that url as what the user wants to see and all should be fine...

**Comment by crevillo** [comment]
hello. here is my investigation about this. 
I could reproduce a similar case, not exactly the same, with an eZ Platform Demo installation from master branch and a short modification.

It looks the symfony Exception listener sets that last requested uri session var when an Access Denied Exception happens, right? 

Well, now please have a look to the pagelayout.html.twig template of ezplatform demo and the subrequest that is made to show the main menu https://github.com/ezsystems/ezplatform-demo/blob/master/app/Resources/views/pagelayout.html.twig#L19

https://github.com/ezsystems/ezplatform-demo/blob/master/src/AppBundle/Controller/MenuController.php#L69

Leaving is at it's now, there's no redirection problem.  For instance, you can go to `ezplatform-demo/Media`, (for non ez platform users, this is an url that by default has no access for anonymous users). You'll be redirected to the login page, you enter your credentials and you'll be redirected to that `/Media` as expected. Ok.

But now, lest's modify that action a bit adding something like 
`$this->locationService->loadLocation(5);` in the  `getChildNodesAction`. For non ez users, that will try to load a location which the anon user has no access. In brief, it will throw an Access Denied exception. 
So, when this acces denied exception is thrown from https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/Repository/LocationService.php#L203, the symfony Exception Listener will try to set that url as the last visited one. And in that case, that url is something like
```
http://local.ezdemo.prod/_fragment?_hash=9wI%2BHT841%2F45aExntZybAUIVqjLSlINIKHDUZp35syE%3D&_path=template%3Dpagelayout_header.html.twig%26serialized_siteaccess%3DO%253A38%253A%2522eZ%255CPublish%255CCore%255CMVC%255CSymfony%255CSiteAccess%2522%253A3%253A%257Bs%253A4%253A%2522name%2522%253Bs%253A4%253A%2522site%2522%253Bs%253A12%253A%2522matchingType%2522%253Bs%253A7%253A%2522default%2522%253Bs%253A7%253A%2522matcher%2522%253BN%253B%257D%26_format%3Dhtml%26_locale%3Den_GB%26_controller%3Dapp.controller.menu%253AgetChildNodesAction
```

That url is set as the last request one. So, if you again go to `/Media` you'll be redirected to login but this time, when login, you'll be redirect to a similar url to the one i've just posted. 

So, i don't really know who to blame here :). As some of you have already said i also think overriding that symfony ExceptionListener is a way to go. We could probably think that symfony could avoid this if it only takes in account masterRequest to be saved as last session requested uri... Could be an option? 

What still i can't reproduce from this ezplatform-demo installation is the case @gggeek is having where   you're redirected to the /_fos_user_context_hash. Maybe first question could be... how i can force a call to that from ez publish or from a symfony controller?

**Comment by dbu** [comment]
i don't know why sub requests are considered in the access exception listener feature. but that is out of scope for this bundle. its either something that should be set up correctly to work as expected or a symfony limitation that ez should work around. i am no expert in this - in our projects we usually have just authorized and anonymous users, and simply do permission checks in the twig template to not include restricted content when the user is not logged in. i guess this pattern does not work if you have more complicated permission schemes, as you don't want to mirror all permission details in your twig templates...

as for the original issue with requesting the hash url, i still think its a misconfiguration of the firewall.

**Comment by dbu** [comment]
nothing against custom listeners. but its not in the domain of this
library/bundle. it could be a cookbook entry in the symfony docs. or a
configuration option on symfony itself - i would try to ask there or at
ez, because it seems like a problem people will run into whenever doing
restricted logins.

**Comment by gggeek** [comment]
ok, will bring this issue to Sf :-)

**Comment by gggeek** [comment]
Back with more findings:

in one of the applications where I am having problems, the whole application is protected by a firewall, except for a couple of urls and for a couple of hostnames (the app shows very different sites based on hostname).

The firewall config is basically the following:
`access_control:
    - { path: ^/, roles: IS_AUTHENTICATED_ANONYMOUSLY, host: ^test\.com$ }
    - { path: ^/free-access/page, roles: IS_AUTHENTICATED_ANONYMOUSLY }
    - { path: ^/_fragment, roles: [IS_AUTHENTICATED_ANONYMOUSLY, ROLE_USER] }
    - { path: ^/, roles: ROLE_USER }
`

The problem here seems to be that _/free-access/page_ can be accessed by anon users ONLY when the http_cache is off. When it is on, requests for that land on redirection to /login.

I now have enabled http_cache and debugging at the same time, and from looking at the event logs, it seems that the 1st "problem" is that requests for _fos_user_context_hash get rejected.

I have then added the following line:
` - { path: ^/_fos_user_context_hash, roles: [IS_AUTHENTICATED_ANONYMOUSLY] }`

and the desired url responds fine... most of the time!
Every now and then, based on time, it still answers back to me with the redirection to the login page :-(
It seems that when that happens, it is because some ESI fragments expire...

http headers for 'good' case:
`* Connected to admin.xxx.local (127.0.0.1) port 80 (#0)
> GET /free-access/page HTTP/1.1
> User-Agent: curl/7.38.0
> Host: admin.xxx.local
> Accept: */*
>
< HTTP/1.1 200 OK
< Date: Thu, 30 Jun 2016 18:21:07 GMT
< Server: Apache
< Cache-Control: must-revalidate, no-cache, private
< Expires: Mon, 28 Jul 1997 05:00:00 GMT
< pragma: no-cache
< x-powered-by: eZ Publish Platform
< served-by: admin.xxx.local
< content-language: en-GB
< X-Cache-Debug: 1
< Vary: X-User-Hash,Accept-Encoding
< X-Debug-Token: ff33ea
< X-Debug-Token-Link: /_profiler/ff33ea
< X-Symfony-Cache: GET /free-access/page: miss
< set-cookie: eZSESSID_XXX=2elommoh2bmurjrhudq5vgf0o3; expires=Thu, 30-Jun-2016 19:21:07 GMT; Max-Age=3600; path=/
< Last-Modified: Thu, 30 Jun 2016 18:21:07 GMT
< Transfer-Encoding: chunked
< Content-Type: text/plain;charset=UTF-8
`
and for the 'bad' case:
`* Connected to web1-www.ovumkc.local (127.0.0.1) port 80 (#0)
> GET /free-access/page HTTP/1.1
> User-Agent: curl/7.38.0
> Host: admin.xxxx.local
> Accept: */*
>
< HTTP/1.1 200 OK
< Date: Thu, 30 Jun 2016 18:21:37 GMT
< Server: Apache
< Cache-Control: must-revalidate, no-cache, private
< Expires: Mon, 28 Jul 1997 05:00:00 GMT
< pragma: no-cache
< x-powered-by: eZ Publish Platform
< served-by: admin.xxx.local
< content-language: en-GB
< X-Cache-Debug: 1
< Vary: X-User-Hash,Accept-Encoding
< X-Debug-Token: 11d644
< X-Debug-Token-Link: /_profiler/11d644
< X-Symfony-Cache: GET /free-access/page: miss; GET /_fragment?_hash=...: miss; GET /_fragment?_hash=...: miss; GET /_fragment?_hash=...: miss
< Set-Cookie: eZSESSID_XXX=pde5pk540lm2kisimvjjudfmh0; expires=Thu, 30-Jun-2016 19:21:37 GMT; Max-Age=3600; path=/
< set-cookie: eZSESSID_XXX=pde5pk540lm2kisimvjjudfmh0; expires=Thu, 30-Jun-2016 19:21:36 GMT; Max-Age=3600; path=/
< Content-Length: 51303
< Content-Type: text/html; charset=utf-8
`

My questions are:

1. I now have the fos_user_hash route in the same firewall as the rest of the site, but there is a rule which makes it accessible to the general public. Is this correct ? Is it a problem ?

2. I suspect that if I allow anon access to **^/_fragment**, using 
`- { path: ^/_fragment, roles: [IS_AUTHENTICATED_ANONYMOUSLY] }`
my problems might go away. I would then have to make sure that requests for _fragment are blocked by the Apache server, so that only Symfony can generate them

3. in theory, the ^/free-access/page controller does not make use of any esi-includes or other subcontrollers (I am not 100% sure because the eZP legacy kernel is involved...). So why would I see the cache-misses for _fragment when I try to access it ?

**Comment by dbu** [comment]
1. indeed the hash lookup url must always be allowed for all users, its kind of the same logic as the login url. i fixed the doc a bit in #292 and will now update the bundle doc as well.

2. when you fetch the hash and don't accidentally redirect, i don't see why anything would load fragments into the page, unless ezpublish is doing something weird in a response listener. the context hash listener aborts the request and sends a response, no controller or twig template should be reached.

3. you would need to investigate this with the ez people, its out of scope for this library and i am not familiar with how ez works. if there is something going on there, and the fragments also are not available without being logged in, it could explain why you run into problems (but then the problem would exist even when not using the user context hash mechanism at all)

**Comment by gggeek** [comment]
About 2 and 3: maybe the flow is the opposite, and _fragment calls are a bit of a red herring:

- request the url
- get user_hash
- something triggers access denied
- display login page
- which has esi blocks
- try to load them
=> cache miss displayed

which brings me back to the original point... more debugging ahead! :-)

**Comment by gggeek** [comment]
Found the critter! Bug on ezpublish side, which tries to do auth unexpectedly when the url in use is the one of the fos auth bundle (it seems that legacy kernel gets bootstrapped and stores the auth url instead of the good url...)

**Comment by dbu** [comment]
excellent! can we close this issue? and could you link to an issue/PR in the ez repositories for reference if somebody else has a similar problem?

**Comment by gggeek** [comment]
Closing as requested
